File: ddd-basic/shop/src/main/java/basic/shop/catalog/command/domain/ProductId.java
package basic.shop.catalog.command.domain;

import lombok.AllArgsConstructor;
import lombok.Getter;

import javax.persistence.Access;
import javax.persistence.AccessType;
import javax.persistence.Column;
import javax.persistence.Embeddable;
import java.io.Serializable;
import java.util.Objects;

@Embeddable
@Access(AccessType.FIELD)
@Getter
@AllArgsConstructor
public class ProductId implements Serializable {
    @Column(name = "product_id")
    private String id;

    @Override
    public boolean equals(Object o) {
        if(o == this)
            return true;
        if(!(o instanceof ProductId productId))
            return false;
        return Objects.equals(id, productId.id);
    }

    @Override
    public int hashCode() {
        return id == null ? 0 : id.hashCode();
    }
}

File: spring-mvc-with-kotlin/order-api/src/main/kotlin/basic/orderapi/catalog/command/domain/product/value/ProductId.kt
package basic.orderapi.catalog.command.domain.product.value

import java.io.Serializable
import java.util.UUID
import javax.persistence.Access
import javax.persistence.AccessType
import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
@Access(AccessType.FIELD)
class ProductId(
    @Column(name = "product_id")
    var id: String
): Serializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ProductId

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    companion object {
        fun generate(): ProductId {
            var randomNo = UUID.randomUUID().toString()
            return ProductId(randomNo)
        }
    }
}